File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution(object):
    def minWindow(self, s, t):
        n, m = len(s), len(t)
        if n == 0 or m == 0 or m > n: return ""
        freq = [0] * 256
        for c in t:
            freq[ord(c)] += 1

        start, end, _len, count, ans = 0, 0, float('inf'), 0, 0
        while end < n:
            if freq[ord(s[end])] > 0: count += 1
            freq[ord(s[end])] -= 1
            while count == m:
                if _len > end - start + 1:
                    _len = end - start + 1
                    ans = start
                freq[ord(s[start])] += 1
                if freq[ord(s[start])] > 0: count -= 1
                start += 1
            end += 1
        
        return "" if _len == float('inf') else s[ans: ans + _len]
        

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int n = s.length(), m = t.length();
        if(n == 0 || m == 0 || m > n) return "";
        int[] freq = new int[256];

        for(char c : t.toCharArray()) {
            freq[c]++;
        }

        int start = 0, end = 0, len = Integer.MAX_VALUE, ans = 0, count = 0;
        while(end < n) {
            char endChar = s.charAt(end);
            if(freq[endChar] > 0) count++;
            freq[endChar]--;
            while(count == m) {
                char startChar = s.charAt(start);
                if(len > end - start + 1) {
                    len = end - start + 1;
                    ans = start;
                }
                freq[startChar]++;
                if(freq[startChar] > 0) count--;
                start++;
            }
            end++;
        }

        return len == Integer.MAX_VALUE ? "" : s.substring(ans, ans + len);
    }
}